packages/go_router/lib/src/route.dart

```diff
@@ -449,19 +461,332 @@ class ShellRoute extends RouteBase {
+///               GoRoute(
+///                 path: 'details',
+///                 builder: (BuildContext context, GoRouterState state) =>
+///                     const DetailsScreen(label: 'B'),
+///               ),
+///             ],
+///           ),
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// When the [Page] for this route needs to be customized, you need to pass a
+/// function for [pageBuilder]. Note that this page provider doesn't replace
+/// the [builder] function, but instead receives the stateful shell built by
+/// [StatefulShellRoute] (using the builder function) as input. In other words,
+/// you need to specify both when customizing a page. For example:
+///
+/// ```
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     StatefulShellRoute.rootRoutes(
+///       builder: (BuildContext context, GoRouterState state,
+///           Widget navigationContainer) {
+///         return ScaffoldWithNavBar(body: navigationContainer);
+///       },
+///       pageBuilder:
+///           (BuildContext context, GoRouterState state, Widget statefulShell) {
+///         return NoTransitionPage<dynamic>(child: statefulShell);
+///       },
+///       routes: <GoRoute>[
+///         /// The first branch, i.e. root of tab 'A'
+///         GoRoute(
+///           parentNavigatorKey: _tabANavigatorKey,
+///           path: '/a',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'A', detailsPath: '/a/details'),
+///         ),
+///         /// The second branch, i.e. root of tab 'B'
+///         GoRoute(
+///           parentNavigatorKey: _tabBNavigatorKey,
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// To access the current state of this route, to for instance access the
+/// index of the current route branch - use the method
+/// [StatefulShellRoute.of]. For example:
+///
+/// ```
+/// final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example using StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
+  /// Constructs a [StatefulShellRoute] from a list of [ShellRouteBranch], each
+  /// representing a root in a stateful route branch.
+  ///
+  /// A separate [Navigator] will be created for each of the branches, using
+  /// the navigator key specified in [ShellRouteBranch].
+  StatefulShellRoute({
+    required List<ShellRouteBranch> branches,
+    required ShellRouteBuilder builder,
+    ShellRoutePageBuilder? pageBuilder,
```

Comment:
Could these be defined on `ShellRouteBase` instead? ShellRoute also has a `ShellRouteBuilder` called `builder` and a `ShellRoutePageBuilder` called `pageBuilder`.

Reply:
Yes, they could, and they were actually in a previous iteration. However, the `pageBuilder` in `StatefulShellRoute` works a bit different, in that it isn't a replacement for the builder, but used in addition to it. So I wanted to catch that in documentation. But I guess that documentation could be placed on class level on `StatefulShellRoute` instead.